WMIAdventure/frontend/src/js/api/data-models/battle/Deck.js

```diff
@@ -0,0 +1,74 @@
+export class Deck {
+    /** @type [Card] */
+    cards = [];
+
+    constructor(cards) {
+        let initialPosition = 1;
+        for (const card of cards) {
+            card.initialPosition = initialPosition;
+            this.cards.push(card);
+            initialPosition++;
+        }
+    }
+
+    lookupCardOnTop() {
+        if (this.cards.length < 1)
+            return null;
+        return this.cards[0];
+    }
+
+    /**
+     * Returns a card from deck with given id
+     * @param id: {int}
+     * @return Card
+     */
+    lookupCardById(id) {
+        return this.cards.filter(card => card.id === id)[0];
+    }
+
+    /**
+     * Returns a card by its position in deck
+     * @param pos: {int}
+     * @return {Card}
+     */
+    lookupCardByPos(pos) {
+        return this.cards.filter(card => card.initialPosition === pos)[0];
+    }
+
+    getCardIdxById(id) {
+        const card = this.cards.filter(card => card.id === id)[0];
+        return this.cards.indexOf(card);
+    }
+
+    /**
+     * Returns a card that was on top of the deck and enqueues it back.
+     * @returns Card
+     */
+    dequeue() {
+        const card = this.cards.shift()
+        this.cards.push(card)
+        return card;
+    }
+
+    /**
+     * Sets a new order of the deck by a given array of card ids.
+     * @throws RangeError when cardIds length does not match the deck's
+     * @param cardIds {[int]}
+     */
+    reorder(cardIds) {
+        if (cardIds.length !== this.cards.length)
+            throw new RangeError();
+
+        const newDeck = []
+        for (const nextId of cardIds) {
+            const card = this.cards.filter(card => card.id === nextId)[0];
+            newDeck.push(card)
+        }
+        this.cards = newDeck;
+    }
+
+    onTurnEnd() {
+        for (const card of this.cards)
+            card.onTurnEnd();
+    }
```

Comment:
na backendzie to inaczej działa, karta ma updatowane rzeczy typu turns stopped i aktywność buffów tylko przy użyciu danej karty - tutaj powinno być wykonywane `onTurnEnd` tylko na użytej karcie

Reply:
jeśli dobrze pamiętam, to chodziło nam o to, że jak zbufujemy następną kartę, a potem zrobimy randomizację decka, to żeby ten buffik dalej był i czekał sobie na wykonanie tej karty, jak przyjdzie znowu na nią kolej